src/bridge/MigrationNFT.sol

```diff
@@ -43,46 +56,60 @@ contract MigrationNFT is ERC721 {
      */
     function tokenURI(uint256 tokenId) public view virtual override(ERC721) returns (string memory) {
         _requireOwned(tokenId);
-        return tokensURI;
+
+        uint256 level = tokenIdToLevel[tokenId];
+        return string.concat(tokensURIRoot, "/", level.toString());
     }
 
     /**
      * @notice Mint a new NFT for the given user
      * @param txHash the transaction hash to bridge
      */
     function safeMint(bytes32 txHash) public {
-        _mustHaveEnoughNFTsRemaining();
         _mustNotHaveBeenClaimed(txHash);
 
         (address target, uint256 amount,,, bool executed) = migration.proposals(txHash);
 
         _mustBeAnExistingProposal(target);
-        _mustBeAboveMinAmount(amount);
         _mustBeExecuted(executed);
-        _mustNotBeAlreadyAHolder(target);
+        bool alreadyHolder = _mustAlreadyBeHolderOrEnougHoldersRemaining(target);
 
-        uint256 tokenId = nextTokenId++;
-        _markAsClaimed(txHash);
-        _safeMint(target, tokenId);
-    }
+        (uint256[] memory levelsToMint, uint256 nbLevelsToMint) = _computeLevelUps(target, amount);
 
-    /**
-     * @notice Mint a batch of NFTs for all the provided transaction hashes
-     * @param txHashes the transaction hashes to mint NFTs for
-     */
-    function safeMintBatch(bytes32[] memory txHashes) public {
-        for (uint256 i = 0; i < txHashes.length; i++) {
-            safeMint(txHashes[i]);
+        claimed[txHash] = true;
+        if (!alreadyHolder) {
+            individualHolders++;
         }
-    }
 
-    function _markAsClaimed(bytes32 txHash) internal {
-        claimed[txHash] = true;
+        for (uint256 i = 0; i < nbLevelsToMint; i++) {
+            uint256 tokenId = nextTokenId++;
+            tokenIdToLevel[tokenId] = levelsToMint[i] + 1;
+            holderToLevel[target] = levelsToMint[i] + 1;
+            _safeMint(target, tokenId);
+        }
     }
 
-    function _mustHaveEnoughNFTsRemaining() internal view {
-        if (nextTokenId >= maxNFTs) {
-            revert TooManyNFTs();
+    function _computeLevelUps(address target, uint256 amount)
+        internal
+        view
+        returns (uint256[] memory levelsToMint, uint256 nbLevelsToMint)
+    {
+        levelsToMint = new uint256[](levels.length);
+        nbLevelsToMint = 0;
+
+        // We effectively iterate over all the levels the `target` has YET
+        // to qualify for. This expressively skips levels the `target` has
+        // already qualified for.
+        uint256 currentLevel = holderToLevel[target];
+        for (uint256 i = currentLevel; i < levels.length; i++) {
+            if (amount >= levels[i]) {
+                levelsToMint[i - currentLevel] = i;
+                nbLevelsToMint++;
+            }
+        }
```

Comment:
If I'm reading the logic correctly here, an individual would be able to get multiples of tokens of the same rank if they have multiple bridge transactions of the same level. If I'm correct is that what we want or we want to issue only one token for an achieved rank?

Reply:
I think my own answer: `holderToLevel` keeps track of a person's currently expected next level! ;)

Reply:
In the light of my answer above I suggest naming `holderToLevel` as `holderNextLevel` and `currentLevel` as `nextLevel`. This means the loop starts from checking a person's eligibility from their next level up.

Reply:
Yes that's fine, makes it more expressive. And yes indeed we just allocate all soulbounds if you are on overachiever

Reply:
Did some renaming and added documentation comments to clarify